src/control/lib/hostlist/hostlist.go

```diff
@@ -0,0 +1,735 @@
+}
+
+// RangedString returns a string containing a bracketed HostList representation.
+func (hl *HostList) RangedString() string {
+	hl.RLock()
+	defer hl.RUnlock()
+
+	var bld strings.Builder
+	var open bool
+	var next *hostRange
+
+	for i, hr := range hl.ranges {
+		next = nil
+		if i+1 < len(hl.ranges) {
+			next = hl.ranges[i+1]
+		}
+
+		if !open {
+			bld.WriteString(hr.prefix)
+			if hr.count() > 1 || (next != nil && hr.within(next)) {
+				open = true
+				bld.WriteString("[")
+			}
+		}
+
+		bld.WriteString(hr.rangedString())
+
+		if open {
+			if next != nil && hr.within(next) {
+				bld.WriteString(",")
+				continue
+			} else {
+				open = false
+				bld.WriteString("]")
+			}
+		}
+
+		if !open && len(hr.suffix) > 0 {
+			bld.WriteString(hr.suffix)
+		}
+
+		if i < len(hl.ranges)-1 {
+			bld.WriteString(",")
+		}
+	}
+
+	return bld.String()
+}
+
+// DerangedString returns a string containing the hostnames of
+// every host in the HostList, without any bracketing.
+func (hl *HostList) DerangedString() string {
+	hl.RLock()
+	defer hl.RUnlock()
+
+	var bld strings.Builder
+
+	for i, hr := range hl.ranges {
+		bld.WriteString(hr.derangedString())
+		if i < len(hl.ranges)-1 {
+			bld.WriteString(",")
+		}
+	}
+
+	return bld.String()
+}
+
+// Push adds a string representation of hostnames to this HostList.
+func (hl *HostList) Push(stringHosts string) error {
+	other, err := Create(stringHosts)
+	if err != nil {
+		return err
+	}
+
+	return hl.PushList(other)
+}
+
+// PushHost adds a single host to this HostList.
+func (hl *HostList) PushHost(stringHost string) error {
+	hn, err := createHostName(stringHost)
+	if err != nil {
+		return err
+	}
+
+	hl.Lock()
+	defer hl.Unlock()
+
+	if hn.hasNumber {
+		return hl.pushRange(&hostRange{
+			prefix:  hn.prefix,
+			suffix:  hn.suffix,
+			lo:      hn.number,
+			hi:      hn.number,
+			width:   hn.width,
+			isRange: true,
+		})
+	}
+
+	return hl.pushRange(&hostRange{
+		prefix: stringHost,
+	})
+}
+
+func (hl *HostList) pushRange(hr *hostRange) error {
+	if len(hl.ranges) > 0 {
+		tail := hl.ranges[len(hl.ranges)-1]
+		if tail.canAppend(hr) {
+			tail.hi = hr.hi
+			hl.hostCount += hr.count()
+			return nil
+		}
+	}
+
+	hl.ranges = append(hl.ranges, hr)
+	hl.hostCount += hr.count()
+
+	return nil
+}
+
+// PushList adds the supplied HostList onto this HostList.
+func (hl *HostList) PushList(other *HostList) error {
+	hl.Lock()
+	defer hl.Unlock()
+	other.RLock()
+	defer other.RUnlock()
+
+	for _, hr := range other.ranges {
+		if err := hl.pushRange(hr); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func fmtRangeHost(hr *hostRange, num uint) string {
+	return fmt.Sprintf("%s%0*d%s", hr.prefix, hr.width, num, hr.suffix)
+}
+
+// Pop returns the string representation of the last host pushed
+// onto the HostList and then removes it from the HostList. Returns
+// an error if the HostList is empty.
+func (hl *HostList) Pop() (hostName string, err error) {
+	if hl.IsEmpty() {
```

Comment:
I'm still working through this, but I'm wondering if you might need to take the lock here before you check for h1.IsEmpty().  Seems like a small window of opportunity for this to get the wrong result.

Reply:
Yeah, I was just looking at that again. I think I'll change it as you suggest. I found a weird case after fuzz testing where a range was merged when it shouldn't have been. It manifested as an out of bounds panic which wouldn't have happened if I was checking the length of the ranges slice under a lock.